### ISSUE TITLE
String (and NULL) default values get wrapped in additional single quotes when scaffolding from MariaDB

### ISSUE THREAD

**Issue by evertonvps** [PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#994]
### Steps to reproduce

1. Scaffold-DbContext "Server=localhost;User=root;Password=123456;Database=newDb;" Pomelo.EntityFrameworkCore.Mysql -Context MyContext -OutputDir Models -Force
2. Add-migration initial
3. update-database

### The issue
I noticed that after scaffolding the hasdefault property is always being generated with single quotes in excess. With this I can not run update-database on my migrations.
I made a comparison with scaffoding by entityframeworkcore and there is the difference of the excess quotation marks in the default values.
![image](https://user-images.githubusercontent.com/3638953/71086301-4f648000-2178-11ea-8c63-f8697c739ae7.png)

```
Exception message:
Stack trace:
```
.
### Further technical details

MySQL version: MariaDb 10.3 64x
Operating system: win10
Pomelo.EntityFrameworkCore.MySql version: 3.0.0 / 3.0.1
Microsoft.AspNetCore.App version: -

Other details about my project setup:

**Comment by lauxjpn** [comment]
Please post the `CREATE TABLE` statement for the table in question.

**Comment by evertonvps** [comment]
@lauxjpn , The statement is this.
`
CREATE TABLE `centro_custo` (
  `id` int(11) NOT NULL,
  `descricao` varchar(100) NOT NULL,
  `ativo` char(1) NOT NULL DEFAULT 'S' COMMENT 'S-Sim | N-Não',
  `criado_em` timestamp NULL DEFAULT NULL,
  `alterado_em` timestamp NULL DEFAULT NULL,
  `excluido_em` timestamp NULL DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8;`

**Comment by lauxjpn** [comment]
Your `'descricao' varchar(100) NOT NULL,` line does not contain any default value. Therefore there shouldn't be generated *any* `HasDefaultValueSql()` call at all.

Also with the `NOT NULL` clause, it does not allow `NULL` values at all, so MySQL would not let you define a `NULL` value default for it.

Are you sure that this is the actual table you are scaffolding and that this is actual SQL definition of the table before you scaffolded it?

**Comment by evertonvps** [comment]
Sorry man! I ended up sending the wrong table. I was able to reproduce the problem with another table.
`CREATE TABLE `empresa` (
  `id` int(11) NOT NULL,
  `fantasia` varchar(150) DEFAULT NULL,
  `razao_social` varchar(150) NOT NULL,
  `tipo` char(1) NOT NULL DEFAULT 'F' COMMENT 'M-Matriz, F-Filial',
  `cnpj` varchar(14) DEFAULT NULL,
  `inscricao_estadual` varchar(25) DEFAULT NULL,
  `endereco` varchar(255) DEFAULT NULL,
  `bairro` varchar(150) DEFAULT NULL,
  `cidade` varchar(100) DEFAULT NULL,
  `cep` varchar(8) DEFAULT NULL,
  `complemento` varchar(150) DEFAULT NULL,
  `criado_em` timestamp NULL DEFAULT NULL,
  `alterado_em` timestamp NULL DEFAULT NULL,
  `excluido_em` timestamp NULL DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8;`
 **Note the excess of single quotes in the **Tipo** column. when performing scaffolding.**

`                entity.Property(e => e.Tipo)
                    .IsRequired()
                    .HasColumnName("tipo")
                    .HasColumnType("char(1)")
                    .HasDefaultValueSql("'''F'''")
                    .HasComment("M-Matriz, F-Filial")
                    .HasCharSet(Pomelo.EntityFrameworkCore.MySql.Storage.CharSet.Utf8);`
**When I run update-database, throws exception**

PM> update-database
Build started...
Build succeeded.
Applying migration '20191222132848_init'.
Failed executing DbCommand (17ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
CREATE TABLE `empresa` (
    `id` int(11) NOT NULL AUTO_INCREMENT,
    `fantasia` varchar(150) NULL DEFAULT 'NULL',
    `razao_social` varchar(150) NOT NULL,
    `tipo` char(1) NOT NULL DEFAULT '''F''',
    `cnpj` varchar(14) NULL DEFAULT 'NULL',
    `inscricao_estadual` varchar(25) NULL DEFAULT 'NULL',
    `endereco` varchar(255) NULL DEFAULT 'NULL',
    `bairro` varchar(150) NULL DEFAULT 'NULL',
    `cidade` varchar(100) NULL DEFAULT 'NULL',
    `cep` varchar(8) NULL DEFAULT 'NULL',
    `complemento` varchar(150) NULL DEFAULT 'NULL',
    `criado_em` timestamp NULL DEFAULT 'NULL',
    `alterado_em` timestamp NULL DEFAULT 'NULL',
    `excluido_em` timestamp NULL DEFAULT 'NULL',
    CONSTRAINT `PK_empresa` PRIMARY KEY (`id`)
);
MySql.Data.MySqlClient.MySqlException (0x80004005): Invalid default value for 'tipo'
 ---> MySql.Data.MySqlClient.MySqlException (0x80004005): Invalid default value for 'tipo'
   at MySqlConnector.Core.ResultSet.ReadResultSetHeaderAsync(IOBehavior ioBehavior) in C:\projects\mysqlconnector\src\MySqlConnector\Core\ResultSet.cs:line 49
   at MySql.Data.MySqlClient.MySqlDataReader.ActivateResultSet() in C:\projects\mysqlconnector\src\MySqlConnector\MySql.Data.MySqlClient\MySqlDataReader.cs:line 130
   at MySql.Data.MySqlClient.MySqlDataReader.CreateAsync(CommandListPosition commandListPosition, ICommandPayloadCreator payloadCreator, IDictionary`2 cachedProcedures, IMySqlCommand command, CommandBehavior behavior, IOBehavior ioBehavior, CancellationToken cancellationToken) in C:\projects\mysqlconnector\src\MySqlConnector\MySql.Data.MySqlClient\MySqlDataReader.cs:line 391
   at MySqlConnector.Core.CommandExecutor.ExecuteReaderAsync(IReadOnlyList`1 commands, ICommandPayloadCreator payloadCreator, CommandBehavior behavior, IOBehavior ioBehavior, CancellationToken cancellationToken) in C:\projects\mysqlconnector\src\MySqlConnector\Core\CommandExecutor.cs:line 62
   at MySql.Data.MySqlClient.MySqlCommand.ExecuteNonQueryAsync(IOBehavior ioBehavior, CancellationToken cancellationToken) in C:\projects\mysqlconnector\src\MySqlConnector\MySql.Data.MySqlClient\MySqlCommand.cs:line 226
   at MySql.Data.MySqlClient.MySqlCommand.ExecuteNonQuery() in C:\projects\mysqlconnector\src\MySqlConnector\MySql.Data.MySqlClient\MySqlCommand.cs:line 74
   at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteNonQuery(RelationalCommandParameterObject parameterObject)
   at Microsoft.EntityFrameworkCore.Migrations.MigrationCommand.ExecuteNonQuery(IRelationalConnection connection, IReadOnlyDictionary`2 parameterValues)
   at Microsoft.EntityFrameworkCore.Migrations.Internal.MigrationCommandExecutor.ExecuteNonQuery(IEnumerable`1 migrationCommands, IRelationalConnection connection)
   at Microsoft.EntityFrameworkCore.Migrations.Internal.Migrator.Migrate(String targetMigration)
   at Pomelo.EntityFrameworkCore.MySql.Migrations.Internal.MySqlMigrator.Migrate(String targetMigration)
   at Microsoft.EntityFrameworkCore.Design.Internal.MigrationsOperations.UpdateDatabase(String targetMigration, String contextType)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.UpdateDatabaseImpl(String targetMigration, String contextType)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.UpdateDatabase.<>c__DisplayClass0_0.<.ctor>b__0()
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.Execute(Action action)
Invalid default value for 'tipo'

**Comment by lauxjpn** [comment]
@evertonvps Thank you for reporting this issue!

It seems that MariaDB 10.2.7 broke compatibility to MySQL with their `INFORMATION_SCHEMA.COLUMNS.COLUMN_DEFAULT` implementation.

Lets assume the following table:

```sql
CREATE TABLE `MariaDBDefaultValueTest` (
  `Id` int(11) NOT NULL,
  `NonNullableWithoutDefault` varchar(255) NOT NULL,
  `NonNullableWithDefaultValueNull` varchar(255) NOT NULL DEFAULT 'NULL',
  `NullableWithoutDefault` varchar(255) NULL,
  `NullableWithDefaultValueNull` varchar(255) NULL DEFAULT 'NULL',
  `NullableWithDefaultNull` varchar(255) NULL DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;
```

Until MariaDB 10.2.6, values were saved as they are still in MySQL:

COLUMN_NAME | COLUMN_DEFAULT | Comment
-- | -- | --
Id | _null_ | actual DBNull
NonNullableWithoutDefault | _null_ | actual DBNull
NonNullableWithDefaultValueNull | NULL | string not enclosed by single quotes
NullableWithoutDefault | _null_ | actual DBNull
NullableWithDefaultValueNull | NULL | string not enclosed by single quotes
NullableWithDefaultNull | _null_ | actual DBNull

Since MariaDB 10.2.7, they are saved differently:

COLUMN_NAME | COLUMN_DEFAULT | Comment
-- | -- | --
Id | _null_ | actual DBNull
NonNullableWithoutDefault | _null_ | actual DBNull
NonNullableWithDefaultValueNull | 'NULL' | string enclosed by single quotes
NullableWithoutDefault | NULL | string not enclosed by single quotes
NullableWithDefaultValueNull | 'NULL' | string enclosed by single quotes
NullableWithDefaultNull | NULL | string not enclosed by single quotes

There are multiple reports of this issue (e.g. [MDEV-13341](https://jira.mariadb.org/browse/MDEV-13341)), but it seems to have been a deliberate choice (see [MDEV-13132](https://jira.mariadb.org/browse/MDEV-13132)).

---

So this is a divergence between MySQL and MariaDB we need to address.

**Comment by lauxjpn** [comment]
This will be fixed in `3.2.0`.
Until then, you might need to manually fix default values for strings columns in MariaDB 10.2.7+.

For your particular example, it would need to be done like this:

```c#
// Faulty line
.HasDefaultValueSql("'''F'''")

// Correct line
.HasDefaultValueSql("'F'")
```

**Comment by lauxjpn** [comment]
We will actually release the fix sooner in the `3.1.1` release.
